Fix mask handling for flex attention in llama/gemma2/mistral/qwen2

flukeskywalker opened the pull request:
# What does this PR do?

**Issue** 
In gemma2 and mistral modelling code, where the `_update_causal_mask()` methods were not handling the case of the `atn_implementation` being the recently added `"flex_attention"`.

**Main Fix**
For Mistral and Gemma2: Added logic for converting attention mask passed as as a tensor to a BlockMask required for flex attention, or skip any conversion if a BlockMask is given directly by the user. This logic is borrowed from `LlamaModel`.

This change also fixes the issue for Gemma3 and Qwen2 models due to inheritance.

**Additional Changes**
~~- For Llama, Mistral, Gemma2: Slightly modified the previous behavior when flex_attention is enabled. Previously in `LlamaModel`, the behavior was unclear/prone to failure if flex_attention was enabled but the attention mask was neither a Tensor nor a BlockMask. The updated code adds an assertion to explicitly fail in this situation.~~

- Updated type hints and docs to reflect that attention mask may also be a BlockMask.


Fixes #37299 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@ArthurZucker

Commit:
fix BlockMask handling when using flex_attention for llama/mistral/gemma2

Commit:
fix attention_mask types

Rocketknight1 commented:
Hi @flukeskywalker, don't forget to `pip install -e .[quality]` and `make fixup` to get the code quality checks to pass!

flukeskywalker commented:
Ah it looks like the failing checks were repo consistency checks because I updated `LlamaModel.update_causal_mask()` as well as type hints in derived models such as `LlamaForSequenceClassification` etc.,  but this method and the type hints are meant to be copied over to lots of other models. 

I will try to fix by running `make fixup` and removing some type hints that I added that create conflicts downstream due to bring copied from Llama models.

Commit:
revert type hints and fixup

Rocketknight1 commented:
CI looks green now, cc @arthurzucker for flexattention review

ArthurZucker reviewed (commented):
LGTM but let's revert the new assert please 😉

Commit:
remove unnecessary assertion

Rocketknight1 commented:
Merging since the review comments have been addressed!
